packages/kbn-optimizer/src/node/node_auto_tranpilation.ts

```diff
@@ -135,56 +73,82 @@ export function registerNodeAutoTranspilation() {
   }
   installed = true;
 
+  const cacheLog = process.env.DEBUG_NODE_TRANSPILER_CACHE
+    ? Fs.createWriteStream(Path.resolve(REPO_ROOT, 'node_auto_transpilation_cache.log'))
+    : undefined;
+
+  const cacheDir = Path.resolve(
+    REPO_ROOT,
+    'data/node_auto_transpilation_cache_v5',
+    UPSTREAM_BRANCH
+  );
+
+  /**
+   * @babel/register uses a JSON encoded copy of the config + babel.version
+   * as the cache key for files, so we do something similar but we don't need
+   * a unique cache key for every file as our config isn't different for
+   * different files (by design). Instead we determine a unique prefix and
+   * automatically prepend all paths with the prefix to create cache keys
+   */
+
   const cache = new Cache({
+    dir: cacheDir,
+    log: cacheLog,
     pathRoot: REPO_ROOT,
-    dir: Path.resolve(REPO_ROOT, 'data/node_auto_transpilation_cache_v4', UPSTREAM_BRANCH),
-    prefix: determineCachePrefix(),
-    log: process.env.DEBUG_NODE_TRANSPILER_CACHE
-      ? Fs.createWriteStream(Path.resolve(REPO_ROOT, 'node_auto_transpilation_cache.log'), {
-          flags: 'a',
+    prefix: Crypto.createHash('sha256')
+      .update(
+        JSON.stringify({
+          synthPkgMapHash: readHashOfPackageMap(),
```

Comment:
Do we invalidate the entire cache if any package is added or removed?

Reply:
Yes, the import statements in every file could resolve differently based on a change in the package map.